### PR TITLE
Adding new blog posts and viewing them.

### PR DESCRIPTION
**Pull request by saadyousafarbi**
This **PR** contains functionality that will allow user to add a new blog post through **`Redux blog post API`**. The following utilities have been used to make this possible:

* **[redux-form](https://redux-form.com/)** : For better management of `form state` in **Redux** (version **5**)
* **[axios](https://github.com/axios/axios)** and **[redux-promise](https://www.npmjs.com/package/redux-promise)** to handle asynchronous request and promises from requests respectively.

### Main Page
main_page

### Creating a new Post page
add_post

### CODE REVIEW

File: src/components/PostNew.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import React, { Component } from 'react';
+import { reduxForm } from 'redux-form';
+import { createPost } from '../actions/index';
+
+class PostNew extends Component {
```
Comment (zubair-arbi) [inline src/components/PostNew.js]: Rename the file to `post.js`
Rename the class to `Post`

File: src/components/PostNew.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    render() {
+        const { fields: { title, categories, content }, handleSubmit } =this.props;
+        return (
+            <form onSubmit={handleSubmit(this.props.createPost)}>
+                <h3> Create A New Post</h3>
+                <div>
+                    <label> Title </label>
+                    <input type='text' {...title.input}/>
+                    <div>
+                        {title.error}
```
Comment (zubair-arbi) [inline src/components/PostNew.js]: How this error will be populated?

File: src/components/PostNew.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            </form>
+        );
+    }
+}
+
+// connect: first argument is mapStateToProps, 2nd is mapDispatchToProps
+// reduxForm: 1st argument is config, 2nd is mapStateToProps, 3rd is mapDispatchToProps
+
+export default reduxForm({
+    form: 'PostNewForm',
```
Comment (zubair-arbi) [inline src/components/PostNew.js]: `PostForm`

File: src/routes.js [hunk 4]
```diff
@@ -3,10 +3,11 @@ import { Route, IndexRoute } from 'react-router';
 
 import App from './components/App';
 import PostIndex from './components/PostIndex';
-
+import PostNew from './components/PostNew';
 
 export default (
     <Route path='/' component={App}>
         <IndexRoute component={PostIndex} />
+        <Route path='post/new' component={PostNew} />
```
Comment (zubair-arbi) [inline src/routes.js]: path should be `post/add/`

File: src/actions/index.js [hunk 5]
```diff
@@ -14,3 +15,13 @@ export function fetchPosts() {
     };
 
 }
+
+
```
Comment (zubair-arbi) [inline src/actions/index.js]: nit: remove this extra newline

File: src/actions/index.js [hunk 6]
```diff
@@ -14,3 +15,13 @@ export function fetchPosts() {
...
+export function createPost(props) {
+    const request = axios.post(`${ROOT_URL}/posts${API_KEY}`, props);
```
Comment (zubair-arbi) [inline src/actions/index.js]: My understanding is that the axios has already fetched the response here. We should name the const `request` to `new_post`. Also have you considered to use `catch`  function here incase server throws error?

File: src/components/PostIndex.js [hunk 7]
```diff
@@ -10,6 +11,11 @@ class PostIndex extends Component {
     render() {
         return (
             <div>
+                <div>
+                    <Link to='/post/new'>
```
Comment (zubair-arbi) [inline src/components/PostIndex.js]: `/post/add/`

File: src/actions/index.js (hunk 6 above)
Comment (saadyousafarbi) [reply src/actions/index.js]: Yes, you are right. I have not considered a server error thus far. I will address these changes.

**Comment by saadyousafarbi** [comment]
I have shifted from **`redux-promise`** middleware to **`redux-thunk`** middleware as it is easier to `catch` server errors on `axios` requests as **`redux-thunk`** uses `dispatch` and returns a function instead of an object.

**Comment by zubair-arbi** [comment]
@verdan please have a look on this PR too.

**Comment by zubair-arbi** [comment]
@saadyousafarbi I suggest merging this PR for now and address comments later (if any).

**Comment by verdan** [comment]
Looks good to me. Nice Work !

**Pull request merged**